apps/astarte_trigger_engine/lib/astarte_trigger_engine/policy/policy.ex

```diff
@@ -0,0 +1,154 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule Astarte.TriggerEngine.Policy do
+  use GenServer
+  require Logger
+
+  alias Astarte.Core.Triggers.Policy.Handler
+  alias Astarte.Core.Triggers.Policy
+  alias Astarte.TriggerEngine.Config
+  # TODO use the ExRabbitPool.RabbitMQ adapter when it will have a `nack` function
+  alias AMQP.Basic
+
+  @consumer Config.events_consumer!()
+
+  # API
+  def start_link(args \\ []) do
+    realm_name = Keyword.fetch!(args, :realm_name)
+    policy = Keyword.fetch!(args, :policy)
+
+    case GenServer.start_link(__MODULE__, args, name: via_tuple(realm_name, policy.name)) do
+      {:ok, pid} ->
+        {:ok, pid}
+
+      {:error, {:already_started, pid}} ->
+        # Already started, we don't care
+        {:ok, pid}
+    end
+  end
+
+  def handle_event(pid, channel, payload, meta) do
+    _ =
+      Logger.debug(
+        "policy process #{inspect(pid)} got event, payload: #{inspect(payload)},  meta: #{
+          inspect(meta)
+        }",
+        tag: "policy_handle_event"
+      )
+
+    GenServer.cast(pid, {:handle_event, channel, payload, meta})
+  end
+
+  # Server callbacks
+
+  def init(args) do
+    policy = Keyword.get(args, :policy)
+    state = %{policy: policy, retry_map: %{}}
+    {:ok, state}
+  end
+
+  def handle_cast(
+        {:handle_event, chan, payload, meta},
+        %{policy: policy, retry_map: retry_map} = _state
+      ) do
+    {headers, _other_meta} = Map.pop(meta, :headers, [])
+    headers_map = amqp_headers_to_map(headers)
+
+    verify_event_consumed = @consumer.consume(payload, headers_map)
+    retry_map = Map.update(retry_map, meta.message_id, 1, fn value -> value + 1 end)
+
+    _ =
+      Logger.debug(
+        "Handling event #{meta.message_id}, this is the #{Map.get(retry_map, meta.message_id)}-th time"
+      )
+
+    case verify_event_consumed do
+      # All was ok
+      :ok ->
+        Basic.ack(chan, meta.delivery_tag)
+        retry_map = Map.delete(retry_map, meta.message_id)
+        {:noreply, %{policy: policy, retry_map: retry_map}}
+
+      {:http_error, status_code} ->
+        maybe_requeue_message(chan, meta, status_code, policy, retry_map)
+
+      {:error, :trigger_not_found} ->
+        do_not_requeue_message(chan, meta, policy, retry_map)
+
+      {:error, reason} ->
+        _ =
+          Logger.warn("Error #{reason} while processing event #{meta.message_id}",
+            tag: "event_consume_error"
+          )
+
+        maybe_requeue_message(chan, meta, nil, policy, retry_map)
```

Comment:
As it is, this either gets always requeue or not since `nil` can never be in the handled `error_codes`

Reply:
Idea: don't handle unexpected errors and just let it crash. This way at least we fail loudly.